File: LeetCode Java/783. Minimum Distance Between BST Nodes.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
 class Solution {
     int minDiff = Integer.MAX_VALUE;
     int val = 0;
     public int minDiffInBST(TreeNode root) {
         dfs(root);
         return minDiff;
     }
     public void dfs(TreeNode root){
         if(root == null) return;
         
         dfs(root.left);
         if(val == 0){
             val = root.val;
         }
         else{
             minDiff = Math.min(minDiff, Math.abs(val - root.val));
             val = root.val;
         }
         dfs(root.right);
     }
 }

File: LeetCode C++/Problems/783. Minimum Distance Between BST Nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void Inorder(TreeNode* curr, int& pre, int& diff) {
        if (curr == nullptr || diff == 1) {
            return;
        }
        Inorder(curr->left, pre, diff);
        if (pre == INT_MIN) {
            pre = curr->val;
        } else {
            diff = min(curr->val - pre, diff);
            pre = curr->val;
        }
        Inorder(curr->right, pre, diff);
    }

    int minDiffInBST(TreeNode* root) {
        int pre = INT_MIN, diff = INT_MAX;
        Inorder(root, pre, diff);
        return diff;
    }
};
